### PR TITLE
Palindrome Checker

### PR DESCRIPTION
**Pull request by Hasan-Z**
### **Summary:**
This PR implements a function to check if a given string is a palindrome, addressing the feature requested in issue #69 .

### **Behavior:**
- The function `is_palindrome` takes an input string and returns `True` if it is a palindrome (the string reads the same forward and backward), and `False` otherwise.
- The function is case-insensitive and ignores spaces.

### **Files Modified:**
- **palindrome_checker.py**: Contains the main function to check for palindromes.
  - Added a module docstring.
  - Added the `is_palindrome` function to check if a string is a palindrome.
  - The function ignores spaces and is case-insensitive.
  - Includes defensive assertions and handling for invalid inputs (raises `ValueError` if the input is not a string).
  - Includes doctests for common cases, including edge cases like empty strings and strings with spaces.
- **test_palindrome_checker.py**: Contains unit tests for the `is_palindrome` function.
  - Added a test class `TestPalindromeChecker` in PascalCase.
  - Each test function has a clear name and docstring.
  - Tests include cases for palindromes, non-palindromes, and edge cases (e.g., empty strings and strings with spaces).
  - All tests pass successfully.

### **Link to Issue:**
Linked to Issue #69 .

### **Checklist**:
- [x] The file name describes the function's behavior (`palindrome_checker.py`).
- [x] There is a module docstring in the function file.
- [x] The test file's name matches the function file name (`test_palindrome_checker.py`).
- [x] There is a module docstring in the test file.
- [x] The test class has a helpful name in PascalCase (`TestPalindromeChecker`).
- [x] The test class has a docstring.
- [x] All unit tests pass.
- [x] Tests include defensive assertions and boundary cases.
- [x] The function's name and file name match and describe its behavior (`is_palindrome` in `palindrome_checker.py`).
- [x] Function has correct type annotations.
- [x] Code passes formatting checks, and no commented-out code or print statements exist.

### **Strategy**:
- The `is_palindrome` function normalizes the string by converting it to lowercase and removing spaces. It then compares the string with its reverse.
- The function raises a `ValueError` if the input is not a string.
- The solution uses Python string slicing to reverse the string and simple string comparisons.

### **Testing:**
- The function is tested with various inputs, including:
  - Palindromes (e.g., "madam", "A man a plan a canal Panama").
  - Non-palindromes (e.g., "hello", "world").
  - Edge cases like an empty string and strings with spaces (e.g., "race car").
  - Handling non-string inputs (raises `ValueError`).

### **Implementation Details**:
- The function checks if the string is a palindrome by normalizing it (removing spaces and converting to lowercase) and comparing it to its reverse.
- It includes input validation and raises a `ValueError` for invalid inputs.
- The code is optimized for simplicity and clarity.

### CODE REVIEW

**Comment by Ibrahim-Elmisbah** [comment]
Hi @WuorBhang, I hope you're doing well! 
I noticed that you merged the branch before I completed my review, which contradicts the workflow that we agreed to. I assigned the review to myself four hours ago and then put the issue under review. 
I hope this does not happen again to ensure regular workflow.

thanks!

**Comment by WuorBhang** [comment]
> Hi @WuorBhang, I hope you're doing well! I noticed that you merged the branch before I completed my review, which contradicts the workflow that we agreed to. I assigned the review to myself four hours ago and then put the issue under review. I hope this does not happen again to ensure regular workflow.
> 
> thanks!

I was fixing a conflict between the branch on settings.json file that we haven't edited and solve the conflict since I found out that all your codes have been approved by someone else and I just merged it after solving the conflict.

Sorry I maybe wrong for merging the branch

File: solutions/palindrome_checker.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+"""
+This module contains a function to check if a string is a palindrome.
+
+A palindrome is a string that reads the same forward and backward, ignoring spaces, punctuation, and capitalization.
+"""
+
+
+def is_palindrome(s: str) -> bool:
```
[the remark this answers is not in the document]
Comment (Hasan-Z) [reply solutions/palindrome_checker.py:8]: matched function name with the file name

File: solutions/tests/test_palindrome_checker.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        Test edge cases including single characters and strings with special characters.
+        """
+        self.assertTrue(is_palindrome("a"))
+        self.assertTrue(is_palindrome(" "))
+        self.assertTrue(is_palindrome("!!"))
+        self.assertTrue(is_palindrome("Able , was I saw eLba"))
+
+
+if __name__ == "__main__":
+    unittest.main()
```
[the remark this answers is not in the document]
Comment (Hasan-Z) [reply solutions/tests/test_palindrome_checker.py:46]: done

File: solutions/palindrome_checker.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        "A man, a plan, a canal: Panama",
+        "Racecar",
+        "Hello",
+        "",
+        "12321",
+        "Not a palindrome",
+    ]
+
+    for test in test_strings:
+        print(f"'{test}' is a palindrome: {is_palindrome(test)}")
```
[the remark this answers is not in the document]
Comment (Hasan-Z) [reply solutions/palindrome_checker.py]: done

File: solutions/tests/test_palindrome_checker.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ibrahim-Elmisbah) [reply solutions/tests/test_palindrome_checker.py]: - [x] Unit tests have only one assertion
  - [x] There are tests for defensive assertions

All checked ✓

File: solutions/palindrome_checker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Ibrahim-Elmisbah) [reply solutions/palindrome_checker.py]: - [x] The function's name matches the file name

All checked ✓

File: solutions/palindrome_checker.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    Returns:
+        bool: True if the input string is a palindrome, False otherwise.
+
+    Example:
+        >>> is_palindrome("A man, a plan, a canal: Panama")
+        True
+        >>> is_palindrome("Hello")
+        False
+    """
```
[the remark this answers is not in the document]
Comment (Ibrahim-Elmisbah) [reply solutions/palindrome_checker.py]: - [x] Each assumption about an argument is checked with an assertion
- [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

All checked ✓

**Review by Ibrahim-Elmisbah — APPROVED** [review]
